akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -115,6 +132,75 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with GenericOut
+    private var wasClosed: Boolean = false
+    private var outstandingStreamWindow: Int = Http2Protocol.InitialWindowSize // adapt if we negotiate greater sizes by settings
+    outlet.setHandler(this)
+
+    def onPull(): Unit = dispatchNextChunk()
+    override def onDownstreamFinish(): Unit = {
+      multiplexer.pushControlFrame(RstStreamFrame(streamId, ErrorCode.CANCEL))
+      incomingStreams -= streamId
+    }
+
+    def onDataFrame(data: DataFrame): Option[IncomingStreamState] = {
+      if (data.endStream) wasClosed = true
+
+      outstandingStreamWindow -= data.sizeInWindow
+      if (outstandingStreamWindow < 0) {
+        multiplexer.pushControlFrame(RstStreamFrame(streamId, ErrorCode.FLOW_CONTROL_ERROR))
+        // also close response delivery if that has already started
+        multiplexer.cancelSubStream(streamId)
+        Some(Closed)
+      } else {
+        buffer ++= data.payload
+        log.debug(s"Received DATA ${data.sizeInWindow} for stream [$streamId], remaining window space now $outstandingStreamWindow, buffered: ${buffer.size}")
+        dispatchNextChunk()
+        None // don't change state
+      }
+    }
+    def onRstStreamFrame(rst: RstStreamFrame): Unit = {
+      outlet.fail(new PeerClosedStreamException(rst.streamId, rst.errorCode))
+      buffer = ByteString.empty
+      wasClosed = true
+    }
+
+    private def dispatchNextChunk(): Unit = {
+      if (buffer.nonEmpty && outlet.isAvailable) {
+        val dataSize = buffer.size min settings.requestEntityChunkSize
+        outlet.push(buffer.take(dataSize))
+        buffer = buffer.drop(dataSize)
+
+        totalBufferedData -= dataSize
+
+        log.debug(s"Dispatched chunk of $dataSize for stream [$streamId], remaining window space now $outstandingStreamWindow, buffered: ${buffer.size}")
+        updateWindows()
+      }
+      if (buffer.isEmpty && wasClosed) outlet.complete()
+    }
+
+    private def updateWindows(): Unit = {
+      val IncomingFlowController.WindowIncrements(connectionLevel, streamLevel) = flowController.onStreamDataDispatched(
+        outstandingConnectionLevelWindow, totalBufferedData,
+        outstandingStreamWindow, buffer.size)
+
+      if (connectionLevel > 0) {
+        multiplexer.pushControlFrame(WindowUpdateFrame(Http2Protocol.NoStreamId, connectionLevel))
+        outstandingConnectionLevelWindow += connectionLevel
+      }
+      if (streamLevel > 0 && !wasClosed /* No reason to update window on closed streams */ ) {
+        multiplexer.pushControlFrame(WindowUpdateFrame(streamId, streamLevel))
+        outstandingStreamWindow += streamLevel
+      }
+
+      log.debug(
+        s"adjusting con-level window by $connectionLevel, stream-level window by $streamLevel," +
+          s"remaining window space now $outstandingStreamWindow, buffered: ${buffer.size}," +
+          s"remaining connection window space now $outstandingConnectionLevelWindow, total buffered: $totalBufferedData")
```

Comment:
Bikeshedding (from looking at review comments): missing space after commas when the string is constructed like this.